### PR TITLE
Added ability to select the next client with the mouse in the preview

### PR DESCRIPTION
**Pull request by wyv3rn**
I'm kinda new to Lua, so any improvement suggestions are welcome. But basically the change should do it's job, i.e. select the appropriate clients as soon as the mouse enters the widget in the preview.

### CODE REVIEW

**Comment by berlam** [comment]
Hi @wyv3rn,

thank you for the contribution. I just want you to know, that I have seen the PR but am currently heavily involved in customer projects. After a quick view on the changes I am not completely sure, why a secondary table is created, when the requirement is to make the previews clickable. Will try to understand it on the weekend, when I have more time for open-source work!

**Comment by wyv3rn** [comment]
Hey @berlam,

at the position where I added the mouse handler, we know the index i of the preview widget in the leftRightTab. For setting the correct altTabIndex, we need the index j of the corresponding client in the altTabTable. However, i and j are not identical due to the fact that the clients are (re)arranged in two halfs for the leftRightTab (as far as I understood: right half starts with the most recent client in the history and left half ends with the oldest client). That's why I introduced the table leftRightTabToAltTabIndex for mapping i to j. An alternative coud be to recalculate j for the mouse handler the same way as it is done for setting up the previews.

For my fork, I'm gonna change it to an identical mapping soon anyway, as I think that's more intuitive (i.e. most recent client at first position, oldest client at last position).

File: init.lua [hunk 1]
```diff
@@ -60,6 +60,29 @@ local source = string.sub(debug.getinfo(1,'S').source, 2)
 local path = string.sub(source, 1, string.find(source, "/[^/]*$"))
 local noicon = path .. "noicon.png"
 
+local function cycle(altTabTable, altTabIndex, dir)
```
Comment (berlam) [inline init.lua]: Looks like you moved the function up. Please revert this change and focus on your enhancement.

File: init.lua [hunk 2]
```diff
@@ -82,6 +105,7 @@ local function preview()
 
    -- create a list that holds the clients to preview, from left to right
    local leftRightTab = {}
+   local leftRightTabToAltTabIndex = {} -- save mapping from leftRightTab to altTabTable as well
```
Comment (berlam) [inline init.lua]: Using a second table is not necessary from my point of view, as we can trigger preview activation with a mouse event as you already did a few lines below with the event handler.

File: init.lua [hunk 3]
```diff
@@ -60,6 +60,29 @@ local source = string.sub(debug.getinfo(1,'S').source, 2)
 local path = string.sub(source, 1, string.find(source, "/[^/]*$"))
 local noicon = path .. "noicon.png"
 
+local function cycle(altTabTable, altTabIndex, dir)
+   -- Switch to next client
+   altTabIndex = altTabIndex + dir
+   while altTabIndex > #altTabTable do
```
Comment (berlam) [inline init.lua]: I am not seeing the improvement here. Please explain your change.

File: init.lua (hunk 1 above)
Comment (wyv3rn) [inline init.lua]: If I move the cycle function to its original position, I either have to 
* make it global instead of local
* or foward declare it, i.e.
`local cycle` at the position where I put it and `cycle = function(...)` and the function body at the original position

I prefer the second variant. If you agree, then I will implement it this way.

File: init.lua (hunk 3 above)
Comment (wyv3rn) [inline init.lua]: The orignal "wrap around" only works if `dir = 1` or `dir = -1`). As this is not the case when arbitrary clients are selected with the mouse, I implemented a proper wrap around (similar to modular arithmetic, but off by 1 due to Lua indices starting at 1). A `if` instead of the `while` loop should be enough as well for the moment, but to be honest, the while loop doesn't hurt here (and makes the implementation more generic for further features where `dir` maybe gets very big).

File: init.lua (hunk 2 above)
Comment (wyv3rn) [inline init.lua]: > as you already did a few lines below with the event handler.

I don't get this one. I'm using the table for the event handler. So without the table, the whole implementation would have to be changed (see my first comment some days ago). I don't see a problem with an additional table which only holds some integers?

File: init.lua (hunk 3 above)
Comment (berlam) [inline init.lua]: I am fine with that. Thank you for the explanation!

File: init.lua (hunk 2 above)
Comment (berlam) [inline init.lua]: Overall the codebase is already on the way to "unmaintainable", that's why I am a little bit harsh to feature additions and especially new state holders. A major refactoring is on my list but not as high on my priority list as other projects. I am looking forward to see some improvements in your fork. Looks promising!

The referenced code block was
```lua
preview_widgets[i]:connect_signal("mouse::enter", function()
         altTabIndex = cycle(altTabTable, altTabIndex, leftRightTabToAltTabIndex[i] - altTabIndex)
end)
```
but using the index here alone is not possible as you already mentioned above. 

For this one I would say that the user does more profit from the feature than from cleaner code.
Nevertheless, thank you for taking time to explain your motivation.

File: init.lua (hunk 1 above)
Comment (berlam) [inline init.lua]: I see the problem with the invocation of `cycle(...)` in the `switch(...)` block. Well done!

**Comment by berlam** [comment]
Thank you for answering my questions and arguing. I will merge your PR.

**Comment by wyv3rn** [comment]
You're welcome. By the way:

> For my fork, I'm gonna change it to an identical mapping soon anyway, as I think that's more intuitive (i.e. most recent client at first position, oldest client at last position).

I have done that in the meantime. This change makes both the `leftRightTab` and the additional `leftRightTabToAltTabIndex` table I introduced unnecessary.
